### PR TITLE
Cardinality of filter Optimization

### PR DESCRIPTION
**Pull request by kaikalur**
Please make sure your submission complies with our [Development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [Formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), and [Commit Message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests) guidelines.

Fill in the release notes towards the bottom of the PR description.
See [Release Notes Guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) for details.

```
== RELEASE NOTES ==

General Changes
* We now optimize the following patterns
  * CARDINALITY(FILTER(a, x -> f(x)))   ===>  REDUCE(a, cast(0 as bigint), (s,x) -> case when f(x) then s + 1 else s end, s->s)
  * CARDINALITY(FILTER(a, x -> f(x))) > 0   ===> REDUCE(a, false, (s,x)->case when s then s else f(x) end, s -> s)
  * CARDINALITY(FILTER(a, x -> f(x))) = 0   ===> REDUCE(a, false, (s,x)->case when s then s else f(x) end, s -> NOT s)
  * CARDINALITY(FILTER(a, x -> f(x))) <comparison> n   ===> REDUCE(a, cast(0 as bigint), (s,x)->case when s <comparison> n then s case when f(x) s + 1 else s end, s -> s <comparison> n)

It's controlled by session property: simplify_array_operations with default set to true.
* ...

Hive Changes
* ...
* ...
```

If release note is NOT required, use:

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by wenleix** [comment]
Nice idea of expression optimization. @shixuan-fan , @hellium01  wondering if you can help with a first round of review? -- This helps with us to enlarge the reviewer pool and get you familiar with more part of code~

Speaking of commit convention, in general in Presto code based we always do rebase and don't do merge.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+        private Expression simplifyCardinalityOfFilterComparedToZero(
+                NodeLocation location,
+                Expression arrayArg,
+                ComparisonExpression.Operator operator,
+                LambdaArgumentDeclaration origLambdaArgument,
+                Expression origLambdaBody, String newLambdaArgumentName,
+                String finalLambdaArgumentName,
+                LambdaArgumentDeclaration finalLambdaArgument)
+        {
+            // Comparing to zero can be rewrtitten to a boolean.
```
Comment (shixuan-fan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java]: nit: rewritten

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+            // cardinality(filter(a, x -> f(x)) = 0 becomes reduce(a, false, (s, x) -> case when s then s else f(x) end, s -> !s)
```
Comment (shixuan-fan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java]: This example seems wrong. Let's just use the one in PR message
`CARDINALITY(FILTER(a, x -> f(x))) = 0   ===> REDUCE(a, false, (s, x) -> CASE WHEN s THEN s ELSE f(x) END, s -> NOT s)`

I think one example would suffice.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+            // cardinality(filter(a, x -> f(x)) > 0 becomes reduce(a, false, (s, x) -> case when s then s else f(x) end, s -> s)
+            Expression finalLambdaBody = new Identifier(location, finalLambdaArgumentName, false);
+            ImmutableList.Builder<WhenClause> builder = new ImmutableList.Builder<WhenClause>();
```
Comment (shixuan-fan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java]: Let's just name this `whenClauses`.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+                ComparisonExpression.Operator operator,
+                LambdaArgumentDeclaration origLambdaArgument,
+                Expression origLambdaBody, String newLambdaArgumentName,
+                String finalLambdaArgumentName,
+                LambdaArgumentDeclaration finalLambdaArgument)
+        {
+            // Comparing to zero can be rewrtitten to a boolean.
+            // cardinality(filter(a, x -> f(x)) = 0 becomes reduce(a, false, (s, x) -> case when s then s else f(x) end, s -> !s)
+            // cardinality(filter(a, x -> f(x)) > 0 becomes reduce(a, false, (s, x) -> case when s then s else f(x) end, s -> s)
+            Expression finalLambdaBody = new Identifier(location, finalLambdaArgumentName, false);
```
Comment (shixuan-fan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java]: Let's name this `outputFunction` to be consistent with documentation of reduce function.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+            ImmutableList.Builder<WhenClause> builder = new ImmutableList.Builder<WhenClause>();
+            builder.add(new WhenClause(new Identifier(location, newLambdaArgumentName, false), new Identifier(location, newLambdaArgumentName, false)));
+
+            if (operator == ComparisonExpression.Operator.EQUAL) {
+                // negate the result
```
Comment (shixuan-fan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java]: This comment seems unnecessary.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+                case LESS_THAN_OR_EQUAL:
+                    return true;
+            }
+
+            return false;
+        }
+
+        private Expression simplifyCardinalityOfFilterComparedToZero(
+                NodeLocation location,
+                Expression arrayArg,
```
Comment (shixuan-fan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java]: nit: `array`?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+                ComparisonExpression.Operator operator,
+                LambdaArgumentDeclaration origLambdaArgument,
+                Expression origLambdaBody, String newLambdaArgumentName,
```
Comment (shixuan-fan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java]: nit: line break and ditto for variable name

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+                    return true;
+            }
+
+            return false;
+        }
+
+        private Expression simplifyCardinalityOfFilterComparedToZero(
+                NodeLocation location,
+                Expression arrayArg,
+                ComparisonExpression.Operator operator,
```
Comment (shixuan-fan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java]: nit: static import?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+                LambdaArgumentDeclaration origLambdaArgument,
```
Comment (shixuan-fan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java]: In Presto most of the time we use full name, so let's call this `originalLambdaArgument`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+                case NOT_EQUAL:
+                case GREATER_THAN_OR_EQUAL:
+                case LESS_THAN_OR_EQUAL:
+                    return true;
+            }
+
+            return false;
+        }
+
+        private Expression simplifyCardinalityOfFilterComparedToZero(
```
Comment (shixuan-fan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java]: I think it might be better if this method is written in the following layout:

```
- create case statement
- create inputFunction using the above expression
- create outputFunction
- create reduce function
```

Currently it was somewhat twisted so the logic is not as obvious.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+            Expression finalLambdaBody = new Identifier(location, finalLambdaArgumentName, false);
+            ImmutableList.Builder<WhenClause> builder = new ImmutableList.Builder<WhenClause>();
+            builder.add(new WhenClause(new Identifier(location, newLambdaArgumentName, false), new Identifier(location, newLambdaArgumentName, false)));
+
+            if (operator == ComparisonExpression.Operator.EQUAL) {
+                // negate the result
+                finalLambdaBody = new NotExpression(finalLambdaBody);
+            }
+
+            LambdaArgumentDeclaration newLambdaArgument = new LambdaArgumentDeclaration(new Identifier(location, newLambdaArgumentName, false));
```
Comment (shixuan-fan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java]: This is `inputFunction` I believe?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+                    location,
+                    QualifiedName.of("reduce"),
+                    ImmutableList.of(
+                            arrayArg,
+                            new BooleanLiteral("false"),
+                            newLambda,
+                            finalLambda));
+        }
+
+        private Expression simplifyCardinalityOfFilter(
```
Comment (shixuan-fan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java]: Let's also revamp this method to be in the following layout to facilitate review process, also, let's also rename the variables following the comments in `simplifyCardinalityOfFilterComparedToZero`:
```
- create cast (not required if it could be inlined)
- create case statement
- create input function
- create output function
- create reduce function
```

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 13]
```diff
@@ -8247,6 +8248,53 @@ public void testRightJoinWithEmptyProbeSide()
                 "WITH small_part AS (SELECT * FROM part WHERE name = 'a') SELECT lineitem.orderkey FROM small_part RIGHT JOIN lineitem ON  small_part.partkey = lineitem.partkey");
     }
 
+    @Test
+    public void CardinalityOfFilterSimplification()
```
Comment (shixuan-fan) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: I think for tests you would have to turn on the session property.

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java [hunk 14]
```diff
@@ -664,6 +665,11 @@ public SystemSessionProperties(
                         OPTIMIZED_REPARTITIONING_ENABLED,
                         "Experimental: Use optimized repartitioning",
                         featuresConfig.isOptimizedRepartitioningEnabled(),
+                        false),
+                booleanProperty(
+                        SIMPLIFY_ARRAY_OPERATIONS,
+                        "Simplify and optimize array operations.",
+                        false,
```
Comment (shixuan-fan) [inline presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java]: Let's also add a corresponding config in `featuresConfig` and use the configured value as default.

**Comment by shixuan-fan** [comment]
Let's rename the commit title and PR title to "Optimize cardinality of filter expression" and put benchmark results in commit message. Also, is there a benchmark class (like `BenchmarkArrayAggregation`) that you are using?

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java [hunk 15]
```diff
@@ -13,77 +13,32 @@
...
-import com.facebook.presto.bytecode.BytecodeBlock;
-import com.facebook.presto.bytecode.BytecodeNode;
-import com.facebook.presto.bytecode.ClassDefinition;
-import com.facebook.presto.bytecode.FieldDefinition;
-import com.facebook.presto.bytecode.MethodDefinition;
-import com.facebook.presto.bytecode.Parameter;
-import com.facebook.presto.bytecode.ParameterizedType;
-import com.facebook.presto.bytecode.Scope;
-import com.facebook.presto.bytecode.Variable;
+import com.facebook.presto.bytecode.*;
```
Comment (hellium01) [inline presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java]: Can you rebase and squash the unrelated commits?

Comment (hellium01) [reply presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java]: Also, please follow: https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines
If you are using IntelliJ (which I highly recommend), definitely import the style file from here: https://github.com/airlift/codestyle.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperations.java [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import static com.facebook.presto.sql.planner.iterative.rule.SimplifyArrayOperationsRewriter.simplifyArrayOperations;
+
+public class SimplifyArrayOperations
+        extends ExpressionRewriteRuleSet
```
Comment (hellium01) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperations.java]: Unfortunately `ExpressionRewriteRuleSet` will be deprecated :) (we really should annotate it with `@Deprecated`), can you also port the change to `RowExpressionRewriteRuleSet`?

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java [hunk 17]
```diff
@@ -664,6 +665,11 @@ public SystemSessionProperties(
                         OPTIMIZED_REPARTITIONING_ENABLED,
                         "Experimental: Use optimized repartitioning",
                         featuresConfig.isOptimizedRepartitioningEnabled(),
+                        false),
+                booleanProperty(
+                        SIMPLIFY_ARRAY_OPERATIONS,
```
Comment (hellium01) [inline presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java]: `SIMPLIFY_ARRAY_OPERATIONS` probably is a little generic? probably `OPTIMIZE_FILTER_CARDINALITY` will be more specific for user?

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java [hunk 18]
```diff
@@ -664,6 +665,11 @@ public SystemSessionProperties(
...
+                        "Simplify and optimize array operations.",
```
Comment (hellium01) [inline presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java]: Probably indicate it will optimize `cardinality(filter(a, x->f(x))` expression in the session description will be more clear to user?

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java (hunk 14 above)
Comment (hellium01) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java]: Though I would argue code change to make it default to true would be easier than rollout an config, it is always safer to have an off switch.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java [hunk 19]
```diff
@@ -0,0 +1,246 @@
...
+
+            return expression instanceof LongLiteral && ((LongLiteral) expression).getValue() == 0;
+        }
+
+        private static boolean isCardinalityOfFilter(Expression expression)
+        {
+            if (expression instanceof FunctionCall) {
+                FunctionCall functionCall = (FunctionCall) expression;
+                return functionCall.getName().toString().equals("cardinality") &&
+                        functionCall.getArguments().size() == 1 &&
```
[the remark this answers is not in the document]
Comment (hellium01) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java]: ditto to this. `cardinality` should always has single argument, if not, we should throw.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java [hunk 20]
```diff
@@ -0,0 +1,246 @@
...
+            FunctionCall filter = (FunctionCall) cardinalityOfFilter.getArguments().get(0);
+            Expression arrayArg = treeRewriter.defaultRewrite(filter.getArguments().get(0), context);
+            LambdaExpression origLambda = (LambdaExpression) filter.getArguments().get(1);
+            LambdaArgumentDeclaration origLambdaArgument = origLambda.getArguments().get(0);
+            Expression origLambdaBody = origLambda.getBody();
+            NodeLocation location = filter.getLocation().orElse(new NodeLocation(0, 0));
+
+            // New lambda arguments
+            // TODO(viswanadha): Fix it to get a unique name that doesn't clash with existing ones.
+            String newLambdaArgumentName = origLambdaArgument.getName().getValue() + "__1__";
```
Comment (hellium01) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java]: You can use `VariableAllocator` to make sure there is no confict.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java [hunk 21]
```diff
@@ -0,0 +1,246 @@
...
+                Void context,
+                ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            // Rewrite CARDINALITY(FILTER(arr, x -> f(x))) as REDUCE(arr, cast(0 as bigint), (s, x) -> case when f(x) then s + 1 else s, s -> s)
+            FunctionCall filter = (FunctionCall) cardinalityOfFilter.getArguments().get(0);
+            Expression arrayArg = treeRewriter.defaultRewrite(filter.getArguments().get(0), context);
+            LambdaExpression origLambda = (LambdaExpression) filter.getArguments().get(1);
+            LambdaArgumentDeclaration origLambdaArgument = origLambda.getArguments().get(0);
+            Expression origLambdaBody = origLambda.getBody();
+            NodeLocation location = filter.getLocation().orElse(new NodeLocation(0, 0));
```
Comment (hellium01) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java]: We don't have to worry about this if we are operating on `RowExpression` :)

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java (hunk 13 above)
Comment (hellium01) [reply presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: I would recommend add separate test similar to `TestSimplifyExpressions` to make sure the expression converted to the desired expression.

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 22]
```diff
@@ -8247,6 +8247,53 @@ public void testRightJoinWithEmptyProbeSide()
...
+                "SELECT cardinality(filter(array_agg(orderkey), x -> x % 2 = 0)) " +
+                "FROM orders " +
+                "WHERE orderkey < 20");
+
+        assertEquals(result.getRowCount(), 1);
+        assertEquals(result.getMaterializedRows().get(0).getField(0), 3L);
+    }
+
+    @Test
+    public void CardinalityOfFilterComparisonSimplification()
```
Comment (hellium01) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: If we have optimized `cardinality(filter(array_agg(orderkey), x -> x < 3))`, would optimizing `cardinality(filter(array_agg(orderkey), x -> x < 3)) = 2` be an overkill?  `=2` seems to be very low cost operations. 

Also, it might make sense to add some benchmark similar to `BenchmarkArrayFilter`, or even do a comparison run from Tpch query runner and add the result as part of commit message.

**Comment by hellium01** [comment]
From another perspective, the reason user will write `cardinality(filter(...))` might be because we lack function do:
```
any(a, predicate)
count_on(a, predicate)
...
```
thus probably another solution is to create such functions?

**Comment by kaikalur** [comment]
Yes that will be a next step but we have several major pipelines using this construct so this will have immediate benefit.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (kaikalur) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java]: I don't like throwing from the optimizer. This is just opportunistic optimization and be defensive and have the exact pattern that you want to match,

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java (hunk 10 above)
Comment (kaikalur) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java]: Yeah I was thinking about it but it is too much code bloat which I dislike very much. I'm not actually not a fan of this kind of rewriting. It should be string/unparse based.

Comment (kaikalur) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java]: I think it's pretty obvious and I put some comments. So I will leave it like this for now.

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java (hunk 13 above)
Comment (kaikalur) [reply presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: The session property is on by default. Don't we use the default configuration/properties for testing?

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java (hunk 14 above)
Comment (kaikalur) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java]: Oops that was a mistake. I want to keep it true by default. Fixed it.

**Comment by kaikalur** [comment]
> Let's rename the commit title and PR title to "Optimize cardinality of filter expression" and put benchmark results in commit message. Also, is there a benchmark class (like `BenchmarkArrayAggregation`) that you are using?

This is a tricky one to measure because it's not  an operator. We can add a benchmark test that loops through this operation a thousand times but not sure.

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java (hunk 22 above)
Comment (kaikalur) [reply presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: The main thing here is to avoid creating the intermediate array.

**Comment by wenleix** [comment]
@kaikalur : The first optimization rule is definitely good to avoid unnecessary array construction.  I am wondering if the second and third optimization rule is necessary? , i.e. 
```
  * CARDINALITY(FILTER(a, x -> f(x))) > 0   ===> REDUCE(a, false, (s,x)->case when s then s else f(x) end, s -> s)
  * CARDINALITY(FILTER(a, x -> f(x))) = 0   ===> REDUCE(a, false, (s,x)->case when s then s else f(x) end, s -> NOT s)
```

In my opinion we should just backport `any_match` from PrestoSQL: https://prestosql.io/docs/current/functions/array.html#any_match  . 

And I am not sure if I understand the fourth optimization rule will be helpful, i.e. 
```
  * CARDINALITY(FILTER(a, x -> f(x))) <comparison> n   ===> REDUCE(a, cast(0 as bigint), (s,x)->case when s <comparison> n then s case when f(x) s + 1 else s end, s -> s <comparison> n)
```

With the first optimization rule, we already avoid unnecessary array construction. And even with this rule we still need to go through the whole array. And my intuition is the overhead of `reduce` function itself will dominate and for now we probably don't need to optimize for this "+1" operation? :)

**Comment by kaikalur** [comment]
> In my opinion we should just backport `any_match` from PrestoSQL: https://prestosql.io/docs/current/functions/array.html#any_match .

Technically, anymatch works only for > 0, for = 0 it doesn't short circuit to return false on the first match.

**Comment by kaikalur** [comment]
> nd my intuition is the overhead of `reduce` function itself will dominate and for now we probably don't need to optimize for this "+1" operation? :)

It will stop on the first time the comparison succeeds so if you have a large array, it will benefit

**Comment by wenleix** [comment]
`all_match()`, `any_match()`, and `none_match()` are being backported in https://github.com/prestodb/presto/pull/13734

**Comment by wenleix** [comment]
@kaikalur :
> Technically, anymatch works only for > 0, for = 0 it doesn't short circuit to return false on the first match.

We have `any_match`, `all_match` and `none_match` 😄 . I believe it should cover this case ~

**Comment by kaikalur** [comment]
> @kaikalur :
> 
> > Technically, anymatch works only for > 0, for = 0 it doesn't short circuit to return false on the first match.
> 
> We have `any_match`, `all_match` and `none_match` 😄 . I believe it should cover this case ~

Too much code, yuck! I would rather have a better short-circuiting reducer. But may be I will back port them for now.

**Comment by kaikalur** [comment]
> > @kaikalur :
> > > Technically, anymatch works only for > 0, for = 0 it doesn't short circuit to return false on the first match.
> > 
> > 
> > We have `any_match`, `all_match` and `none_match` 😄 . I believe it should cover this case ~
> 
> Too much code, yuck! I would rather have a better short-circuiting reducer. But may be I will back port them for now.

And also NonMatch just calls AnyMatch! I'm not backporting that. It's not a performant implementation.

**Comment by rongrong** [comment]
We should also add a SQL function for this one!

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java (hunk 17 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java]: Alternative question to ask: Do you want this flag to guard all future optimization rules that would be added to arrays?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 23]
```diff
@@ -256,8 +257,12 @@ public PlanOptimizers(
 
         PlanOptimizer predicatePushDown = new StatsRecordingPlanOptimizer(optimizerStats, new PredicatePushDown(metadata, sqlParser));
         PlanOptimizer rowExpressionPredicatePushDown = new StatsRecordingPlanOptimizer(optimizerStats, new RowExpressionPredicatePushDown(metadata, sqlParser));
-
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: Nits: Remove this change

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java]: Theoretically you don't need to check that here. Function resolution would have failed if arguments size is not 1. Agree with @hellium01's earlier point that this should be added to `RowExpressionRewriteRuleSet`. Also when you do that, please use proper API (add `isCardinalityFunction` `isFilterFunction` to `StandardFunctionResolution`) rather than relying on name (`CallExpression`'s name is only used for display purpose)

File: presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java (hunk 15 above)
Comment (kaikalur) [reply presto-main/src/main/java/com/facebook/presto/sql/gen/PageFunctionCompiler.java]: I thought I had the airift code style. I will doublecheck

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java (hunk 17 above)
Comment (kaikalur) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java]: > Alternative question to ask: Do you want this flag to guard all future optimization rules that would be added to arrays?

Yeah - I was thinking about that. It's a tricky balance between having too many flags and controlling what is enabled. Since most of these optimizations are opportunistic and hopefully don't need to be turned off frequently, I thought one is just enough. But i can change it if there is a strong reason.

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java (hunk 22 above)
Comment (kaikalur) [reply presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: > If we have optimized `cardinality(filter(array_agg(orderkey), x -> x < 3))`, would optimizing `cardinality(filter(array_agg(orderkey), x -> x < 3)) = 2` be an overkill? `=2` seems to be very low cost operations.
> 
> Also, it might make sense to add some benchmark similar to `BenchmarkArrayFilter`, or even do a comparison run from Tpch query runner and add the result as part of commit message.

The short-circuiting helps when you have a lot of medium-size arrays

**Comment by kaikalur** [comment]
Regarding doing it for RowExpressions - I would say no. The idea is to move these rewrites up the stack to just after parse/resolve and eventually they will be string/template based ones - parse -resovle-(unparse(rewrite)-parse-resolve)*  so people get to see the query structure to do more interesting. I'm doing this one for now because it's widely used pattern.

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java (hunk 18 above)
Comment (kaikalur) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java]: Most users won't even understand what that means. This flag is only for any unforeseen bugs.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java (hunk 8 above)
Comment (kaikalur) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java]: It's surprising that our code formatter doesn't do these things automatically. I will fix it.
